### ISSUE TITLE
Real part of complex expMinusOne is bad when e^{-x} ~ cos y

### ISSUE THREAD

**Issue by riastradh** [apple/swift-numerics#158]
For example, in binary32:

x = 1.09590280055999755859375
y = 1.230000019073486328125

The real part of e^{x + iy} - 1, computed in high precision with Sollya and rounded to binary32, is

-3.4550861727211668039672076702117919921875e-8,

but the expression `fma(expm1(x),cos(y), -versin(y))` as used at https://github.com/apple/swift-numerics/blob/c26e60c6ba6e545cf8d3048a3cee554058f82db1/Sources/ComplexModule/ElementaryFunctions.swift#L117 gives (rounding each subexpression to binary32 in Sollya)

-6.9330411633927724324166774749755859375e-8,

in which every digit is wrong.

The problem is catastrophic cancellation when e^x cos y ~ 1.  Factoring it to compute the addition with FMA still trips over catastrophic cancellation because (e^x - 1) cos y ~ versin y if e^x cos y ~ 1.  Same goes for the other obvious factorings as e^x - 1 - e^x versin t, or e^{x + log cos y} - 1 = e^{x + log(1 - versin y)} - 1, although they do worse when x >>> 1 because the low resolution of floating-point values around y ~ acos(e^{-x}) gives no more opportunity for cancellation once y = fl(pi/2).

I don't see any clear way to compute the error in either term short of computing the transcendental functions in extended precision.  I guess that might be easy for binary32, but in binary64 you'd presumably have to do the transcendental functions in double-Dekker arithmetic.  Maybe there's a clever algebraic trick to avoid the subtraction altogether, but I haven't thought of one.

---

Nit: There appears to be a typo in the comment at: https://github.com/apple/swift-numerics/blob/c26e60c6ba6e545cf8d3048a3cee554058f82db1/Sources/ComplexModule/ElementaryFunctions.swift#L86

It reads

```
exp(x) cosMinuxOne(y) + expMinusOne(y)
```

but it should read

```
exp(x) cosMinusOne(y) + expMinusOne(x)
```

**Comment by stephentyrone** [comment]
Yup. This is a problem that no one has a good solution to yet. I have a few ideas, but haven't had time to sort through them.

As you suggest, the cosm1 factorization is strictly better than the expm1 factorization (it at least gets an _absolute_ error bound for this case, which is why I'm using it).

**Comment by riastradh** [comment]
> > short of computing the transcendental functions in extended precision
> 
> There's not even an easy proof that this suffices--how close can (x,y) be to exp(-x) = cos(y)? This is non-trivial, and you need to know in order to bound the precision required.

I didn't think very hard about how doing it in extended precision would actually work; mostly I just wanted an excuse to say 'double-Dekker arithmetic'.

**Comment by NevinBR** [comment]
I have some thoughts on this, not sure if it’ll help though.

First, if `cos(y)` is negative or zero then we know the real part of the result will be at most -1, hence not near 0, so we can use a simple calculation.

Second, the problem curve approaches `cos(y) = 0` as x increases, so we definitely do not want to use `cosMinusOne(y)` when x is large.

Third, when x is small, the problem curve approaches `(exp(x), cos(y)) = (1, 1)`, so both `expMinusOne(x)` and `cosMinusOne(y)` could plausibly be useful. One way to utilize both is to notice that `a*b - 1 = (a-1)*(b-1) + (a-1) + (b-1)`. That means we can go:

```
let e = .expMinusOne(x)
let c = .cosMinusOne(y)
let u = c*e + c + e        // Or similar with augmented sum / fma / etc.
```

*****

An entirely different approach would be to look at the distance between `(x, y)` and the problem curve. This distance could be measured horizontally, vertically, or perhaps even along the gradient of `cos(y) - exp(-x)`. Horizontal is easiest:

When `cos(y)` is positive, the horizontal distance from the curve to the point is given by `d = x + log(cos(y))`. Of course, that involves catastrophic cancellation near the curve, so maybe it doesn’t buy us much.

Ignoring that issue, if `cos(y)` is close to 1 we might rewrite as `d = x + log(onePlus: cosMinusOne(y))`. In either case, the result we want is just `u = expMinusOne(d)`. I’m not sure there’s an easy way around the cancellation issue though, and when `cos(y)` is close to 1 we’d probably rather use the previous approach anyway.

**Comment by stephentyrone** [comment]
> Second, the problem curve approaches cos(y) = 0 as x increases, so cosMinusOne(y) is actively harmful in that region for large x.

Well, no, because the alternative factorization (`exp(x) cosMinuxOne(y) + expMinusOne(x)`) has terms that get arbitrarily large, while the terms in the factorization we're using stay bounded near 1. So it's not what we want, but it's far better than the only alternative we have available currently.

**Comment by NevinBR** [comment]
My point is that for large enough x, using `expMinusOne` doesn’t gain any precision, and if we are also close to the curve then using `cosMinusOne` actually loses precision. In that situation neither factorization is helpful, so the naive calculation `u = exp(x) * cos(y) - 1` should be just as good and maybe better.

In other places, like when `(x, y)` is close to `(0, 0)`, there is something to gain by using both `expMinusOne(x)` and `cosMinusOne(y)`, so perhaps the `a*b + a + b` dual factorization might outperform either of the single factorizations.

Of course for `z` *extremely* close to `0 + 0i`, the nearest representable value to `expMinusOne(z)` is simply `z`. I don’t know if it helps to special-case that, or just let it fall out naturally from the calculation.

Basically, what I’m suggesting is to partition the plane into some number of regions, and use whichever approach is most accurate in each.

**Comment by stephentyrone** [comment]
Right, `exp(x) * cos(y) - 1` is about as good on that region, but not better. The expMinusOne alternative is always strictly worse. So there's no reason to use either of those on any region of the plane. You're always at least as well off using what we have now, and often better off.

In an additive expression with cancellation, the error comes only from the rounding of the terms that are cancelling. If you look at what we use currently:
```
expMinusOne(x) cos(y) + cosMinusOne(y)
```
Assuming a good-quality host math library, expMinusOne, cos, and cosMinusOne are all computed with small relative error (ideally sub-ulp, a couple of ulp at worst). So we are evaluating an expression of the form a(1+δ₁) + b(1+δ₂) where a and b are the exact real values of (e^x-1)cos(y) and cos(y)-1, respectively and the deltas are bounded by a small multiple of epsilon. Because we're interested in the region close to the critical curve, the addition is computed exactly, so the result is precisely Re(exp(z)-1) + δ₁a + δ₂b, and so the absolute error is bounded by (|a|+|b|)kε for some small k.

Note that this analysis holds for _all_ of the expressions you consider, except the "dual" factorization; the way we minimize the error bound is to minimize `|a|` and `|b|`. Among the alternatives considered, the one in use minimizes them for _all_ input values, not just in some region. There's a small note to make for exp(x)cos(y) - 1, because 1 is exact and therefore does not contribute to the error, but that buys you at most a bit or two in practice in this case, hardly worth the added complexity when you still have essentially no good bits in the result (it _would_ be likely to be worth using in an arbitrary-precision implementation, however).

**Comment by stephentyrone** [comment]
Forgot to address z very nearly 0 😀

In this case, expMinusOne(x) cos(y) + cosMinusOne(y) becomes x*1 + y^2/2 = x, so there's no loss of accuracy in just using the expression we have.

**Comment by NevinBR** [comment]
> In an additive expression with cancellation, the error comes only from the rounding of the terms that are cancelling. If you look at what we use currently:
> ```
> expMinusOne(x) cos(y) + cosMinusOne(y)
> ```
> Assuming a good-quality host math library, expMinusOne, cos, and cosMinusOne are all computed with small relative error (ideally sub-ulp, a couple of ulp at worst). So we are evaluating an expression of the form a(1+δ₁) + b(1+δ₂) where a and b are the exact real values of (e^x-1)cos(y) and cos(y)-1, respectively and the deltas are bounded by a small multiple of epsilon. Because we're interested in the region close to the critical curve, the addition is computed exactly, so the result is precisely Re(exp(z)-1) + δ₁a + δ₂b, and so the absolute error is bounded by (|a|+|b|)kε for some small k.

I’ll trust your expertise here, but it still seems strange to me that we would intentionally calculate `cosMinusOne` when `cos` is close the zero.

**Comment by stephentyrone** [comment]
It's no less accurate than the alternatives we have (as explained), not significantly slower, and yields considerable simplicity in the implementation. What's not to like?

**Comment by NevinBR** [comment]
Several years later, I realize there’s another possible factorization of `e^x cos(y) - 1`, by bringing the cosine into the exponent as `e^(x + ln(cos(y))) - 1`. Of course this only applies when `cos(y) > 0`, and it doesn’t actually solve the catastrophic cancellation, but it’s still potentially interesting.

Notably, we can write it as `expm1(x + log1p(cosm1(y)))`, though I don’t know how to compare the approximation error of this to the other factorizations. I expect it will be similar because of the cancellation.

The important thing is that if we can compute `u = x + ln(cos(y))` to high accuracy, then we can use `expMinusOne(u)` for the final result. And that’s nice because now we only need to compute one series in one variable to high precision, instead of two series in different variables.

Moreover, we can recognize `ln(cos(y))` as an antiderivative of `-tan(y)`, so its Taylor series can be found by integrating that of `tan(y)` termwise. If `y` is much smaller than 1 this converges reasonably quickly, but if not it can take much longer.

If we wanted to use an approach like this, say as a fallback when we recognize the problem case is occurring, there are various ways to improve the convergence. We could recenter the Taylor series (eg. around π/4), we could apply some series acceleration techniques, or we could use a different method of evaluating the integral of `tan(y)` more accurately.

It’s also possible to bring the `x` inside the integral, either as a rectangle or a triangle, or perhaps some other shape that more closely approximates `tan`, though I don’t know if that actually helps.

Anyway, I don’t actually have a concrete suggestion at this time. I just wanted to share this alternative factorization, and how it can reduce the amount of arbitrary-precision calculation down to evaluating a single series (or integral) in `y`.
